.NET MAUI eBook - MVVM Toolkit chapter (#34713)
dotnet/docs: docs/architecture/maui/mvvm-community-toolkit-features.md

# MVVM Toolkit Features

[!INCLUDE [download-alert](includes/download-alert.md)]

## MVVM Toolkit

The [Model-View-ViewModel (MVVM)](mvvm.md) pattern is a great structural basis for creating our applications. In this pattern, the ViewModel becomes the backbone of our application as it provides communication to our front-end user interface and backing components. To provide integration with the user interface, we will rely on the ViewModel's properties and commands. As detailed in [Updating views in response to changes in the underlying view model or model](mvvm.md#updating-views-in-response-to-changes-in-the-underlying-view-model-or-model), the `INotifyPropertyChanged` interface on our ViewModel to allows changes to our properties to notify when the value is changed. Implementing all of these features means that our ViewModel can end up becoming very verbose. For example, the following code shows a simple ViewModel with properties that raise changes:

```csharp
public class SampleViewModel : INotifyPropertyChanged
{
    private string _name;
    private int _value;

    public event PropertyChangedEventHandler PropertyChanged;

    public string Name
    {
        get => _name;

>         set => this.SetPropertyValue(ref _name, value);

Reviewer: ```suggestion
        set => SetPropertyValue(ref _name, value);
```

    }

    public int Value
    {
        get => _value;

>         set => this.SetPropertyValue(ref _value, value);

Reviewer: ```suggestion
        set => SetPropertyValue(ref _value, value);
```

    }

>     protected bool SetPropertyValue<T>(ref T storageField, T newValue, [CallerMemberName] string propertyName = "")

Reviewer: ```suggestion
    protected bool SetPropertyValue<T>(
        ref T storageField,
        T newValue,
        [CallerMemberName] string propertyName = "")
```

I'm curious, why does this return `bool` at all? It's not used here, and it made me question its purpose.

    {
        if (Equals(storageField, newValue))
            return false;

        storageField = newValue;

>         this.RaisePropertyChanged(propertyName);

Reviewer: ```suggestion
        RaisePropertyChanged(propertyName);
```

        return true;
    }

    protected virtual void RaisePropertyChanged([CallerMemberName] string propertyName = "")
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
    }
}
```

While some optimizations could be made over time, we will still end up with a fairly verbose set of code for defining our ViewModel. This code can be difficult to maintain and becomes error-prone.

> The MVVM Toolkit (`CommunityToolkit.Mvvm`) package can be used to help address and simplify these common MVVM patterns. The MVVM Toolkit, along with newer features to the .NET language, allows for simplified logic, easy adoption into a project, and runtime independence. The example below shows the same ViewModel using components that come with the MVVM Toolkit:

Reviewer: Perhaps we should include a link to the NuGet package.

Author: There is a note on line 68 with a link to install instructions. I struggled with where to put this. Maybe I move it up higher. Thoughts?

```
> [!NOTE]
> The MVVM Toolkit is provided with the `CommunityToolkit.Mvvm` package. For information on how to add the package to your project, see [Introduction to the MVVM Toolkit](/dotnet/communitytoolkit/mvvm/) on the Microsoft Developer Center.
```

Reviewer: Ah, yes. The first time it's mentioned introduce it as the CommunityToolkit.Mvvm NuGet package, with link. Then after that just as `CommunityToolkit.Mvvm`.

Author: I updated the language there. Let me know if that is what you were thinking.

```csharp
public partial class SampleViewModel : ObservableObject
{
    [ObservableProperty]
    private string _name;

    [ObservableProperty]
    private int _value;
}
```

> [!NOTE]
> The MVVM Toolkit is provided with the `CommunityToolkit.Mvvm` package. For information on how to add the package to your project, see [Introduction to the MVVM Toolkit](/dotnet/communitytoolkit/mvvm/) on the Microsoft Developer Center.

In comparison to the original example, we were able to drastically reduce the overall complexity and simplify the maintainability of our ViewModel. The MVVM Toolkit comes with many pre-built common components and features, such as the `ObservableObject` shown above, that simplifies and standardizes the code that we have throughout the application.

> ## ObservableObject

Reviewer: ```suggestion
## `ObservableObject`
```

The MVVM Toolkit provides `ObservableObject` which is intended for use as the base of our `ViewModel` objects or any object that needs to raise change notifications. It implements `INotifyPropertyChanged` and `INotifyPropertyChanging` along with helper methods for setting properties and raising changes. Below is an example of a standard ViewModel using `ObservableObject`:

```csharp
public class SampleViewModel : ObservableObject
{
    private string _name;
    private int _value;

    public string Name
    {
        get => _name;

>         set => this.SetProperty(ref _name, value);

Reviewer: ```suggestion
        set => SetProperty(ref _name, value);
```

    }

    public int Value
    {
        get => _value;

>         set => this.SetProperty(ref _value, value);

Reviewer: ```suggestion
        set => SetProperty(ref _value, value);
```

    }
}
```

`ObservableObject` handles all of the logic needed for raising change notifications by using the `SetProperty` method in your property setter. If you have a property that returns a `Task<T>`, the `SetPropertyAndNotifyOnCompletion` method can be used to delay publishing a property change until the task has been completed. The methods `OnPropertyChanged` and `OnPropertyChanging` that can also be used for raising property changes where needed in your object.

For more detailed information on `ObservableObject`, see [ObservableObject](/dotnet/communitytoolkit/mvvm/observableobject) in the MVVM Toolkit Developer Center.

> ## RelayCommand and AsyncRelayCommand

Reviewer: ```suggestion
## `RelayCommand` and `AsyncRelayCommand`
```

> Interaction between .NET MAUI controls (e.g. tapping a button or selecting an item from a collection) and the ViewModel is done with the `ICommand` interface. .NET MAUI comes with a default implementation of `ICommand` with the `Command` object. .NET MAUI's `Command` is fairly basic and lacks support for more advanced features, such as supporting asynchronous work and command execution status.

Reviewer: ```suggestion
Interaction between .NET MAUI controls (for example, tapping a button or selecting an item from a collection) and the ViewModel is done with the `ICommand` interface. .NET MAUI comes with a default implementation of `ICommand` with the `Command` object. .NET MAUI's `Command` is fairly basic and lacks support for more advanced features, such as supporting asynchronous work and command execution status.
```

The MVVM Toolkit comes with two commands, `RelayCommand` and `AsyncRelayCommand`. `RelayCommand` is intended for situations where you have synchronous code to execute and has a fairly similar implementation to the .NET MAUI `Command` object.

> [!NOTE]

> > Even though the.NET MAUI `Command` and `RelayCommand` are similar, using `RelayCommand` allows for decoupling your ViewModel from any direct .NET MAUI references. This means that your ViewModel is more portable which can lead to easier reuse across project.

Reviewer: ```suggestion
> Even though the .NET MAUI `Command` and `RelayCommand` are similar, using `RelayCommand` allows for decoupling your ViewModel from any direct .NET MAUI references. This means that your ViewModel is more portable, leading to easier reuse across projects.
```

`AsyncRelayCommand` provides many additional features when working with asynchronous workflows. This is quite common in our ViewModel as we are typically communicating with repositories, APIs, databases, and other systems that utilize `async/await`. The `AsyncRelayCommand` constructor takes in an execution task defined as a `Func<Task>` or a delegate returning `Task` as part of the constructor. While the execution task is running, `AsyncRelayCommand` will monitor the state of the task and provides updates using the `IsRunning` property. The `IsRunning` property can be bound to the UI which helps manage control states such as showing loading with an `ActivityIndicator` or disabling/enabling a control. While the execution task is being executed, the `Cancel` method can be called to attempt cancellation of the execution task, if supported.

> By default, `AsyncRelayCommand` does not allow concurrent execution. This is very helpful in situations where a user could unintentionally tap a control multiple times to execute a long-running or costly operation. During task execution, `AsyncRelayCommand` will automatically call the `CanExecuteChanged` event. In .NET MAUI, controls that support the `Command` and `CommandParameter` properties, such as `Button`, will listen to this event and automatically enable or disable it during execution. This functionality can be overridden by using a custom `canExecute` parameter or setting the `AsyncRelayCommandOptions.AllowConcurrentExecutions` flag in the constructor.

Reviewer: ```suggestion
By default, `AsyncRelayCommand` doesn't allow concurrent execution. This is very helpful in situations where a user could unintentionally tap a control multiple times to execute a long-running or costly operation. During task execution, `AsyncRelayCommand` will automatically call the `CanExecuteChanged` event. In .NET MAUI, controls that support the `Command` and `CommandParameter` properties, such as `Button`, will listen to this event and automatically enable or disable it during execution. This functionality can be overridden by using a custom `canExecute` parameter or setting the `AsyncRelayCommandOptions.AllowConcurrentExecutions` flag in the constructor.
```

For more detailed information on implementing commands, see the section [Implementing commands](mvvm.md#implementing-commands) in the MVVM chapter. Detailed information for the `RelayCommand` and `AsyncRelayCommand` is available in the [Commanding](/dotnet/communitytoolkit/mvvm/relaycommand) of the MVVM Toolkit Developer Center.

## Source Generators

> Using the MVVM Toolkit components out-of-the-box allows us to greatly simplify our ViewModel. The MVVM Toolkit allows us to simplify common code use cases even further by using [Source Generators](/dotnet/csharp/roslyn-sdk/source-generators-overview). The MVVM Toolkit source generators look for specific attributes in our code and can generate wrappers for properties and commands.

Reviewer: ```suggestion
Using the MVVM Toolkit components out-of-the-box allows you to greatly simplify our ViewModel. The MVVM Toolkit allows you to simplify common code use cases even further by using [Source Generators](/dotnet/csharp/roslyn-sdk/source-generators-overview). The MVVM Toolkit source generators look for specific attributes in our code and can generate wrappers for properties and commands.
```

> > [!NOTE]

Reviewer: ```suggestion
> [!IMPORTANT]
```

> The MVVM Toolkit Source Generators generate code that is additive to our existing objects. Because of this, any object that is leveraging a source generator will need to be marked as `partial`.

> The MVVM Toolkit `ObservableProperty` attribute works on classes based-off on `ObservableObject` and will wrap a private variable as a property that generates changes. The following code shows an example of using the `ObservableObject` attribute:

Reviewer: I'm not sure how to change this, but this part of the sentence

```md
works on classes based-off on
```

doesn't make sense to me. Could you please rephrase that?

Author: Yea, it reads weird for sure. On it now.

Author: I updated it, and I think it is better, but it could be more concise. I wasn't sure of a better way to make sure that it was explicit enough to show where it can be used.

```csharp
public partial class SampleViewModel : ObservableObject
{
    [ObservableProperty]
    private string _name;
}
```

> With the `ObservableProperty` attribute applied to the `_name` variable, the source generator will run and generate another partial class with the following code:

Reviewer: ```suggestion
With the `ObservableProperty` attribute applied to the `_name` field, the source generator will run and generate another partial class with the following code:
```

Sure it is a variable, but so are properties, I think the important thing here is specifying that it's a field.

```csharp
partial class SampleViewModel
{
    public string Name
    {
        get => _name;
        set
        {
            if (!EqualityComparer<string>.Default.Equals(_name, value))
            {
                OnNameChanging(value);
                OnPropertyChanging("Name");
                _name = value;
                OnNameChanged(value);
                OnPropertyChanged("Name");
            }
        }
    }
}
```

> The generated `SampleViewModel` has used our private `_name` variable and generated a new property `Name` that implements all of the logic needed for raising change notifications.

Reviewer: ```suggestion
The generated `SampleViewModel` has used the private `_name` field and generated a new `Name` property that implements all of the logic needed for raising change notifications.
```
